// nuclide-remote-search/lib/LocalFindInProjectService.js
var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

/*
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the LICENSE file in
 * the root directory of this source tree.
 */

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

'use babel';

var _require = require("events");

var EventEmitter = _require.EventEmitter;

var scanhandler = require('./scanhandler');
var path = require('path');

var ON_MATCHES_UPDATE = 'update';
var ON_SEARCH_COMPLETED = 'completed';

var LocalFindInProjectService = (function () {
  function LocalFindInProjectService() {
    _classCallCheck(this, LocalFindInProjectService);

    this._requests = 1;
    this._emitter = new EventEmitter();
  }

  _createClass(LocalFindInProjectService, [{
    key: "search",
    value: function search(directory, regex, caseSensitive, subdirs) {
      var _this = this;

      var requestId = this._requests++; // Get a unique number to represent this request.

      // Start the search asynchronously.
      process.nextTick(function () {
        scanhandler.search(directory, regex, function (update) {
          // On update, normalize the paths, and try to pass the update to clients.
          _this._emitter.emit(ON_MATCHES_UPDATE, requestId, {
            filePath: path.join(directory, update.filePath),
            matches: update.matches
          });
        }, caseSensitive, subdirs).then(function (results) {
          // Upon completion of search, emit event.
          _this._emitter.emit(ON_SEARCH_COMPLETED, requestId);
        });
      });

      // Return the request id to the client, without blocking on the search's completion.
      return Promise.resolve(requestId);
    }

    // Subscribe to the completion of searches.
  }, {
    key: "onSearchCompleted",
    value: function onSearchCompleted(callback) {
      var _this2 = this;

      // Add this callback to a listener for completions.
      this._emitter.addListener(ON_SEARCH_COMPLETED, callback);

      // Return a disposable that lets the client unsubscribe.
      return {
        dispose: function dispose() {
          return _this2._emitter.removeListener(ON_SEARCH_COMPLETED, callback);
        }
      };
    }

    // Subscribe to an event triggered whenever new matches are found in a file.
  }, {
    key: "onMatchesUpdate",
    value: function onMatchesUpdate(callback) {
      var _this3 = this;

      // Add this callback to a listener for results.
      this._emitter.addListener(ON_MATCHES_UPDATE, callback);

      // Return a disposable that lets the client unsubscribe.
      return {
        dispose: function dispose() {
          return _this3._emitter.removeListener(ON_MATCHES_UPDATE, callback);
        }
      };
    }
  }]);

  return LocalFindInProjectService;
})();

module.exports = LocalFindInProjectService;

// Store a current request id counter.

// A gobal emitter used by clients receive streaming results.
