// nuclide-ssh-dialog/lib/open-connection.js

/*
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the LICENSE file in
 * the root directory of this source tree.
 */

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { var callNext = step.bind(null, 'next'); var callThrow = step.bind(null, 'throw'); function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(callNext, callThrow); } } callNext(); }); }; }

var url = require('url');
var path = require('path');
var defaultConfig = null;

var dialogPromiseQueue = null;

function openConnectionDialog(props) {
  var _require = require('nuclide-commons');

  var extend = _require.extend;
  var PromiseQueue = _require.PromiseQueue;

  if (!dialogPromiseQueue) {
    dialogPromiseQueue = new PromiseQueue();
  }

  return dialogPromiseQueue.submit(function (resolve, reject) {
    var workspaceEl = atom.views.getView(atom.workspace);
    var hostEl = document.createElement('div');
    workspaceEl.appendChild(hostEl);

    var ConnectionDialog = require('./ConnectionDialog');

    var defaultConnectionSettings = getDefaultConfig().getConnectionDialogDefaultSettings();

    function saveConfig(config) {
      atom.config.set('nuclide.lastConnectionDetails', {
        config: config,
        // Save last official command to detect upgrade.
        lastOfficialRemoteServerCommand: defaultConnectionSettings.remoteServerCommand
      });
    }

    var lastConnectionDetails = atom.config.get('nuclide.lastConnectionDetails') || {};
    var lastConfig = lastConnectionDetails.config || {};
    var remoteServerCommand = defaultConnectionSettings.remoteServerCommand;
    // If there has not been any upgrade and there is a persisted remote server command, use it.
    if (lastConnectionDetails.lastOfficialRemoteServerCommand === defaultConnectionSettings.remoteServerCommand && lastConfig.remoteServerCommand) {
      remoteServerCommand = lastConfig.remoteServerCommand;
    }
    var rememberedDialogSettings = extend.immutableExtend(defaultConnectionSettings, lastConfig);

    var dialogProps = extend.immutableExtend({
      initialUsername: rememberedDialogSettings.username,
      initialServer: rememberedDialogSettings.host,
      initialRemoteServerCommand: remoteServerCommand,
      initialCwd: rememberedDialogSettings.cwd,
      initialSshPort: String(rememberedDialogSettings.sshPort),
      initialPathToPrivateKey: rememberedDialogSettings.pathToPrivateKey,
      initialAuthMethod: rememberedDialogSettings.authMethod,
      onConnect: _asyncToGenerator(function* (connection, config) {
        resolve(connection);
        saveConfig(config);
      }),
      onError: function onError(err, config) {
        resolve( /*connection*/null);
        saveConfig(config);
      },
      onCancel: function onCancel() {
        return resolve( /*connection*/null);
      }
    }, props);
    var React = require('react-for-atom');
    React.render(React.createElement(ConnectionDialog, dialogProps), hostEl);
  });
}

function getDefaultConfig() {
  if (defaultConfig) {
    return defaultConfig;
  }
  try {
    defaultConfig = require('./fb/config');
  } catch (e) {
    defaultConfig = require('./config');
  }
  return defaultConfig;
}

module.exports = { openConnectionDialog: openConnectionDialog };
