// nuclide-debugger-hhvm-proxy/lib/frame.js

/*
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the LICENSE file in
 * the root directory of this source tree.
 */

var _require = require('./utils');

var uriToPath = _require.uriToPath;

/**
 * A dbgp Frame after it is converted from XML to JSON:
 * {
 *   "$":{
 *     "where":"{main}",
 *     "level":"0",
 *     "type":"file",
 *     "filename":"file:///home/peterhal/test/dbgp/test-client.php",
 *     "lineno":"2"
 *   }
 * }
 */

function idOfFrame(frame) {
  // TODO: Mangle in the transactionId of the most recent pause/status.
  return Number(frame.$.level);
}

function functionOfFrame(frame) {
  return frame.$.where;
}

// Returns an absolute path
function fileOfFrame(frame) {
  return uriToPath(frame.$.filename);
}

function locationOfFrame(frame) {
  return {
    // TODO: columnNumber: from cmdbegin/end
    lineNumber: Number(frame.$.lineno) - 1,
    scriptId: fileOfFrame(frame)
  };
}

module.exports = {
  idOfFrame: idOfFrame,
  functionOfFrame: functionOfFrame,
  fileOfFrame: fileOfFrame,
  locationOfFrame: locationOfFrame
};
