// nuclide-atom-helpers/lib/main.js

/*
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the LICENSE file in
 * the root directory of this source tree.
 */

module.exports = Object.defineProperties({}, {
  ScriptBufferedProcess: {
    get: function get() {
      return require('./script-buffered-process');
    },
    configurable: true,
    enumerable: true
  },
  atomEventDebounce: {
    get: function get() {
      return require('./atom-event-debounce');
    },
    configurable: true,
    enumerable: true
  },
  fileTypeClass: {
    get: function get() {
      return require('./file-type-class');
    },
    configurable: true,
    enumerable: true
  },
  goToLocation: {
    get: function get() {
      return require('./go-to-location');
    },
    configurable: true,
    enumerable: true
  },
  closeTabForBuffer: {
    get: function get() {
      return require('./close-tab-buffer');
    },
    configurable: true,
    enumerable: true
  },
  extractWordAtPosition: {
    get: function get() {
      return require('./extract-word-at-position');
    },
    configurable: true,
    enumerable: true
  },
  mouseListenerForTextEditor: {
    get: function get() {
      return require('./mouse-listener-for-text-editor');
    },
    configurable: true,
    enumerable: true
  },
  observeLanguageTextEditors: {
    get: function get() {
      return require('./observe-language-text-editors');
    },
    configurable: true,
    enumerable: true
  },
  observeGrammarForTextEditors: {
    get: function get() {
      return require('./observe-grammar-for-text-editors');
    },
    configurable: true,
    enumerable: true
  },
  sysinfo: {
    get: function get() {
      return require('./sysinfo');
    },
    configurable: true,
    enumerable: true
  }
});
