// nuclide-service-transformer/lib/service-parser.js

/*
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the LICENSE file in
 * the root directory of this source tree.
 */

var _slicedToArray = (function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i['return']) _i['return'](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError('Invalid attempt to destructure non-iterable instance'); } }; })();

var babel = require('babel-core');
var fs = require('fs');

var _require = require('./method-name-parser');

var isEventMethodName = _require.isEventMethodName;

var cache = new Map();

function parseAst(sourceFilePath) {
  var sourceCode = fs.readFileSync(sourceFilePath, 'utf8');

  return babel.transform(sourceCode, {
    blacklist: ['es6.classes', 'flow', 'strict'],
    optional: ['es7.classProperties']
  }).ast;
}

/**
 * Parse service definition file and return an object like:
 * {className: $className, rpcMethodNames: [$methodName, ...], eventMethodNames: [$methodName, ..]}.
 * Keep it sync as it will be called from NuclideServer's constructor.
 */
function parseServiceApiSync(absoluteServiceDefinitionClassFilePath, serviceName) {
  if (cache.has(absoluteServiceDefinitionClassFilePath)) {
    return cache.get(absoluteServiceDefinitionClassFilePath);
  }

  var ast = parseAst(absoluteServiceDefinitionClassFilePath);

  var _ast$program$body$filter = ast.program.body.filter(function (astNode) {
    return astNode.type === 'ClassDeclaration' && astNode.id.name === serviceName;
  });

  var _ast$program$body$filter2 = _slicedToArray(_ast$program$body$filter, 1);

  var classDeclaration = _ast$program$body$filter2[0];

  var methodNames = classDeclaration.body.body
  // Because class bodies in ES7 can contain static property intializers, ensure a part is a
  // "MethodDefinition" before using it. Static properties have type "ClassProperty".
  .filter(function (bodyPart) {
    return bodyPart.type === 'MethodDefinition';
  }).map(function (methodDefinition) {
    return methodDefinition.key.name;
  });
  var rpcMethodNames = methodNames.filter(function (methodName) {
    return !isEventMethodName(methodName);
  });
  var eventMethodNames = methodNames.filter(function (methodName) {
    return isEventMethodName(methodName);
  });

  var serviceStructure = {
    className: classDeclaration.id.name,
    eventMethodNames: eventMethodNames,
    rpcMethodNames: rpcMethodNames
  };

  cache.set(absoluteServiceDefinitionClassFilePath + '$' + serviceName, serviceStructure);

  return serviceStructure;
}

module.exports = parseServiceApiSync;
