// nuclide-panel/lib/PanelController.js

/*
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the LICENSE file in
 * the root directory of this source tree.
 */

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

var React = require('react-for-atom');

var PanelComponent = require('./PanelComponent');

/**
 * Instantiating this class adds it to the UI (even if it's not visible).
 * It currently does this with `atom.workspace.addLeftPanel()` but should
 * support different sides in the future.
 */

var PanelController = (function () {
  function PanelController(childElement, props, state) {
    _classCallCheck(this, PanelController);

    this._hostEl = document.createElement('div');
    // Fill the entire panel with this div so content can also use 100% to fill
    // up the entire panel.
    this._hostEl.style.height = '100%';

    var shouldBeVisible = false;
    var initialLength = null;
    if (state) {
      props.initialLength = state.resizableLength;
      shouldBeVisible = state.isVisible;
    }

    this._component = React.render(React.createElement(
      PanelComponent,
      props,
      childElement
    ), this._hostEl);
    this._panel = atom.workspace.addLeftPanel({ item: this._hostEl, visible: shouldBeVisible });
  }

  _createClass(PanelController, [{
    key: 'destroy',
    value: function destroy() {
      React.unmountComponentAtNode(this._hostEl);
      this._panel.destroy();
    }
  }, {
    key: 'toggle',
    value: function toggle() {
      this.setVisible(!this.isVisible());
    }
  }, {
    key: 'setVisible',
    value: function setVisible(shouldBeVisible) {
      if (shouldBeVisible) {
        this._panel.show();
        this._component.focus();
      } else {
        this._panel.hide();
      }
    }
  }, {
    key: 'isVisible',
    value: function isVisible() {
      return this._panel.isVisible();
    }
  }, {
    key: 'getChildComponent',
    value: function getChildComponent() {
      return this._component.getChildComponent();
    }
  }, {
    key: 'serialize',
    value: function serialize() {
      return {
        isVisible: this.isVisible(),
        resizableLength: this._component.getLength()
      };
    }
  }]);

  return PanelController;
})();

module.exports = PanelController;
