// nuclide-client/lib/service-manager.js

/*
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the LICENSE file in
 * the root directory of this source tree.
 */

var _slicedToArray = (function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i['return']) _i['return'](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError('Invalid attempt to destructure non-iterable instance'); } }; })();

var logger = require('nuclide-logging').getLogger();

var _require = require('nuclide-server/lib/config');

var loadConfigsOfServiceWithServiceFramework = _require.loadConfigsOfServiceWithServiceFramework;

var _require2 = require('nuclide-server/lib/service-manager');

var optionsToString = _require2.optionsToString;

var _require3 = require('nuclide-remote-connection');

var RemoteConnection = _require3.RemoteConnection;

var _require4 = require('nuclide-remote-uri');

var isRemote = _require4.isRemote;
var getHostname = _require4.getHostname;

var serviceConfigs = loadConfigsOfServiceWithServiceFramework();

// A cache stores services in form of '$serviceName@$host:$options' => $serviceObject. A special
// case would be the local service, where the $host will be empty string.
var cachedServices = new Map();

RemoteConnection.onDidCloseRemoteConnection(function (connection) {
  for (var cacheEntry of cachedServices) {
    var _cacheEntry = _slicedToArray(cacheEntry, 2);

    var cacheKey = _cacheEntry[0];
    var serviceInstance = _cacheEntry[1];

    if (serviceInstance._connection === connection) {
      cachedServices['delete'](cacheKey);
    }
  }
});

/**
 * Create or get a cached service with given serviceOptions.
 * @param nuclideUri It could either be either a local path or a remote path in form of
 *    `nuclide:$host:$port/$path`. The function will use the $host from remote path to
 *    create a remote service with given serviceOptions or create a local service if the
 *    uri is local path.
 */
function getServiceByNuclideUri(serviceName) {
  var nuclideUri = arguments.length <= 1 || arguments[1] === undefined ? null : arguments[1];
  var serviceOptions = arguments.length <= 2 || arguments[2] === undefined ? null : arguments[2];

  var hostname = nuclideUri && isRemote(nuclideUri) ? getHostname(nuclideUri) : null;
  return getService(serviceName, hostname, serviceOptions);
}

/**
 * Create or get a cached service with given serviceOptions. If hostname is null or empty string,
 * it returns a local service, otherwise a remote service will be returned. For the same host
 * serviceOptions, the same service instance will be returned.
 */
function getService(serviceName, hostname, serviceOptions) {
  var _serviceConfigs$filter = serviceConfigs.filter(function (config) {
    return config.name === serviceName;
  });

  var _serviceConfigs$filter2 = _slicedToArray(_serviceConfigs$filter, 1);

  var serviceConfig = _serviceConfigs$filter2[0];

  if (!serviceConfig) {
    logger.error('Service %s undefined.', serviceName);
    return null;
  }

  var cacheKey = serviceName + '@' + (hostname ? hostname : '') + ':' + optionsToString(serviceOptions);

  if (cachedServices.has(cacheKey)) {
    return cachedServices.get(cacheKey);
  }

  serviceOptions = serviceOptions || {};

  if (hostname) {
    var serviceInstance = createRemoteService(serviceConfig, hostname, serviceOptions);
  } else {
    var serviceInstance = createLocalService(serviceConfig, serviceOptions);
  }
  cachedServices.set(cacheKey, serviceInstance);

  return serviceInstance;
}

function createRemoteService(serviceConfig, hostname, serviceOptions) {
  var _require5 = require('nuclide-service-transformer');

  var requireRemoteServiceSync = _require5.requireRemoteServiceSync;

  var remoteServiceClass = requireRemoteServiceSync(serviceConfig.definition, serviceConfig.name);
  var remoteConnection = RemoteConnection.getByHostnameAndPath(hostname, null);
  return new remoteServiceClass(remoteConnection, serviceOptions);
}

function createLocalService(serviceConfig, serviceOptions) {
  var serviceClass = require(serviceConfig.implementation);
  return new serviceClass(serviceOptions);
}

module.exports = {
  getService: getService,
  getServiceByNuclideUri: getServiceByNuclideUri
};
