// nuclide-atom-helpers/lib/sysinfo.js

/*
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the LICENSE file in
 * the root directory of this source tree.
 */

// Since nuclide-logging depends on getUnixname, we should use
// console.error in this file instead of nuclide-logging.
/*eslint-disable no-console */

var _slicedToArray = (function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i['return']) _i['return'](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError('Invalid attempt to destructure non-iterable instance'); } }; })();

var determineFlowVersion = _asyncToGenerator(function* () {
  var flowVersion;
  try {
    var pathToFlow = yield getConfigValueAsync('nuclide-flow.pathToFlow')();
    var result = yield asyncExecute(pathToFlow, ['--version'], {});
    flowVersion = result.stdout.trim();
  } catch (e) {
    flowVersion = '';
  }
  return flowVersion;
});

var determineAllSystemInfo = _asyncToGenerator(function* () {
  var _ref = yield Promise.all([getUnixname(), getOSVersion(), getClangVersion(), getFlowVersion()]);

  var _ref2 = _slicedToArray(_ref, 4);

  var userID = _ref2[0];
  var localOsVersion = _ref2[1];
  var localClangVersion = _ref2[2];
  var flowVersion = _ref2[3];

  return {
    version: atom.getVersion(),
    buildNumber: getBuildNumber(),
    smokeBuildNumber: getSmokeBuildNumber(),
    userID: userID,
    unixname: userID,
    osVersion: localOsVersion,
    clangVersion: localClangVersion,
    flowVersion: flowVersion
  };
});

var addSystemInfoPropertiesTo = _asyncToGenerator(function* (data) {
  var allSystemInfo = yield getAllSystemInfo();
  for (var info in allSystemInfo) {
    // we know that this has only its own properties
    data[info] = allSystemInfo[info];
  }
});

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { var callNext = step.bind(null, 'next'); var callThrow = step.bind(null, 'throw'); function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(callNext, callThrow); } } callNext(); }); }; }

var _require = require('nuclide-commons');

var asyncExecute = _require.asyncExecute;
var getConfigValueAsync = _require.getConfigValueAsync;

var path = require('path');
var atomMeta = require(path.join(atom.getLoadSettings().resourcePath, 'package.json'));

var unixname;
var buildNumber = null;
var smokeBuildNumber = null;
var osVersion;
var clangVersion;

function getUnixname() {
  if (!unixname) {
    unixname = asyncExecute('whoami', [], {}).then(function (result) {
      return result.stdout.trim();
    }, function (result) {
      console.error(result.stderr);
      return '';
    });
  }
  return unixname;
}

function getBuildNumber() {
  if (buildNumber === null) {
    buildNumber = atomMeta.buildNumber ? atomMeta.buildNumber : 0;
  }
  return buildNumber.toString();
}

function getSmokeBuildNumber() {
  if (smokeBuildNumber === null) {
    smokeBuildNumber = atomMeta.smokeBuildNumber ? atomMeta.smokeBuildNumber : 0;
  }
  return smokeBuildNumber.toString();
}

function getOSVersion() {
  if (!osVersion) {
    osVersion = asyncExecute('sw_vers', ['-productVersion'], {}).then(function (result) {
      return result.stdout.trim();
    }, function (result) {
      console.error(result.stderr);
      return '';
    });
  }
  return osVersion;
}

function getClangVersion() {
  if (!clangVersion) {
    clangVersion = asyncExecute('clang', ['-v'], {}).then(function (result) {
      // Clang return this info in stderr.
      return result.stderr;
    }, function (result) {
      console.error(result.stderr);
      return '';
    });
  }
  return clangVersion;
}

var flowVersionPromise;

function getFlowVersion() {
  if (!flowVersionPromise) {
    flowVersionPromise = determineFlowVersion();
  }
  return flowVersionPromise;
}

var allSystemInfoPromise;

function getAllSystemInfo() {
  if (!allSystemInfoPromise) {
    allSystemInfoPromise = determineAllSystemInfo();
  }
  return allSystemInfoPromise;
}

module.exports = {
  getUnixname: getUnixname,
  getBuildNumber: getBuildNumber,
  getSmokeBuildNumber: getSmokeBuildNumber,
  getOSVersion: getOSVersion,
  getClangVersion: getClangVersion,
  getFlowVersion: getFlowVersion,
  getAllSystemInfo: getAllSystemInfo,
  addSystemInfoPropertiesTo: addSystemInfoPropertiesTo
};
