// nuclide-debugger-hhvm-proxy/lib/ChromeCallback.js

/*
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the LICENSE file in
 * the root directory of this source tree.
 */

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

var _require = require('./utils');

var log = _require.log;
var logError = _require.logError;

// Sends chrome dev tools JSON messages to a callback.
// The Chrome dev tools protocol is detailed at:
//   https://developer.chrome.com/devtools/docs/protocol/1.1/index

var ChromeCallback = (function () {
  function ChromeCallback(callback) {
    _classCallCheck(this, ChromeCallback);

    this._callback = callback;
  }

  _createClass(ChromeCallback, [{
    key: 'replyWithError',
    value: function replyWithError(id, error) {
      this.replyToCommand(id, {}, error);
    }
  }, {
    key: 'replyToCommand',
    value: function replyToCommand(id, result, error) {
      var value = { id: id, result: result };
      if (error) {
        value.error = error;
      }
      this._sendJsonObject(value);
    }
  }, {
    key: 'sendMethod',
    value: function sendMethod(method, params) {
      this._sendJsonObject(createMessage(method, params));
    }
  }, {
    key: '_sendJsonObject',
    value: function _sendJsonObject(value) {
      var message = JSON.stringify(value);
      if (this._callback) {
        log('Sending JSON: ' + message);
        this._callback(message);
      } else {
        logError('Attempt to send json after dispose: ' + message);
      }
    }
  }, {
    key: 'dispose',
    value: function dispose() {
      this._callback = null;
    }
  }]);

  return ChromeCallback;
})();

function createMessage(method, params) {
  var result = { method: method };
  if (params) {
    result.params = params;
  }
  return result;
}

module.exports = ChromeCallback;
