// nuclide-buck-base/lib/LocalBuckProject.js

/*
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the LICENSE file in
 * the root directory of this source tree.
 */

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; desc = parent = getter = undefined; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { var callNext = step.bind(null, 'next'); var callThrow = step.bind(null, 'throw'); function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(callNext, callThrow); } } callNext(); }); }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _require = require('nuclide-commons');

var asyncExecute = _require.asyncExecute;

var _require2 = require('nuclide-commons');

var fsPromise = _require2.fsPromise;

var logger = require('nuclide-logging').getLogger();
var path = require('path');
var BuckProject = require('./BuckProject');

/**
 * As defined in com.facebook.buck.cli.Command, some of Buck's subcommands are
 * read-only. The read-only commands can be executed in parallel, but the rest
 * must be executed serially.
 *
 * TODO(mbolin): This does not account for the case where the user runs
 * `buck build` from the command line while Nuclide is also trying to build.
 */
var BLOCKING_BUCK_COMMAND_QUEUE_PREFIX = 'buck';

/**
 * Represents a Buck project on disk. All Buck commands for a project should be
 * done through an instance of this class.
 */

var LocalBuckProject = (function (_BuckProject) {
  _inherits(LocalBuckProject, _BuckProject);

  /**
   * @param options.rootPath Absolute path to the directory that contains the
   *     .buckconfig file to configure the project.
   */

  function LocalBuckProject(options) {
    _classCallCheck(this, LocalBuckProject);

    _get(Object.getPrototypeOf(LocalBuckProject.prototype), 'constructor', this).call(this);
    this._rootPath = options.rootPath;
    this._serialQueueName = BLOCKING_BUCK_COMMAND_QUEUE_PREFIX + this._rootPath;
  }

  _createClass(LocalBuckProject, [{
    key: 'getPath',
    value: function getPath() {
      return Promise.resolve(this._rootPath);
    }

    /**
     * This syntax is not supported yet, but the return type is:
     * Promise<{stdout: string; stderr: string; exitCode: number}>
     *
     * @param args Do not include 'buck' as the first argument: it will be added
     *     automatically.
     */
  }, {
    key: '_runBuckCommandFromProjectRoot',
    value: _asyncToGenerator(function* (args) {
      if (global.atom) {
        var pathToBuck = atom.config.get('buck.pathToBuck');
      } else {
        var pathToBuck = 'buck';
      }
      var options = {
        cwd: this._rootPath,
        queueName: this._serialQueueName
      };
      logger.debug('Buck command:', pathToBuck, args, options);
      return asyncExecute(pathToBuck, args, options);
    })
  }, {
    key: 'getOwner',
    value: _asyncToGenerator(function* (filePath) {
      var args = ['audit', 'owner', filePath];
      var result = yield this._runBuckCommandFromProjectRoot(args);
      var stdout = result.stdout;
      var targets = stdout.trim().split('\n');
      return targets;
    })
  }, {
    key: 'build',
    value: function build(buildTargets) {
      return this._build(buildTargets, { install: false, run: false, debug: false });
    }
  }, {
    key: 'install',
    value: function install(buildTargets, run, debug, simulator) {
      return this._build(buildTargets, { install: true, run: run, debug: debug, simulator: simulator });
    }
  }, {
    key: '_build',
    value: _asyncToGenerator(function* (buildTargets, options) {
      if (typeof buildTargets === 'string') {
        buildTargets = [buildTargets];
      }
      var install = options.install;
      var run = options.run;
      var debug = options.debug;
      var simulator = options.simulator;

      var report = yield fsPromise.tempfile({ suffix: '.json' });
      var args = install ? ['install'] : ['build'];
      args = args.concat(['--keep-going', '--build-report', report]);
      if (install) {
        if (run) {
          args.push('--run');
        }
        if (debug) {
          args.push('--wait-for-debugger');
        }
        if (simulator) {
          args.push('--udid');
          args.push(simulator);
        }
      }
      buildTargets.forEach(function (target) {
        return args.push(target);
      });

      try {
        yield this._runBuckCommandFromProjectRoot(args);
      } catch (e) {
        // The build failed. However, because --keep-going was specified, the
        // build report should have still been written unless any of the target
        // args were invalid. We check the existence of the report file to be sure.
        var fileWasWritten = yield fsPromise.exists(report);
        if (!fileWasWritten) {
          throw e;
        }
      }

      try {
        var json = yield fsPromise.readFile(report, { encoding: 'UTF-8' });
        if (!json) {
          throw Error('Report file ' + report + ' for ' + buildTargets + ' was opened, ' + 'but nothing was written.');
        }

        try {
          return JSON.parse(json);
        } catch (e) {
          throw Error('Failed to parse:\n' + json);
        }
      } finally {
        fsPromise.unlink(report);
      }
    })

    /**
     * @param filePath absolute path.
     */
  }, {
    key: 'findTargetsWithReferencedFile',
    value: _asyncToGenerator(function* (filePath, options) {
      var args = ['targets', '--referenced_file', filePath];

      var type = options['type'];
      if (type) {
        args.push('--type');
        type.forEach(function (buildRuleType) {
          return args.push(buildRuleType);
        });
      }

      args.push('--json');

      var result = yield this._runBuckCommandFromProjectRoot(args);

      result.json = JSON.parse(result.stdout || '[]');
      // TODO(mbolin): The target should be a field in the JSON.
      result.targets = result.json.map(function (targetData) {
        return '//' + targetData['buck.base_path'] + ':' + targetData['name'];
      });

      return result;
    })
  }, {
    key: 'listAliases',
    value: _asyncToGenerator(function* () {
      var args = ['audit', 'alias', '--list'];
      var result = yield this._runBuckCommandFromProjectRoot(args);
      var stdout = result.stdout.trim();
      return stdout ? stdout.split('\n') : [];
    })

    /**
     * Currently, if `aliasOrTarget` contains a flavor, this will fail.
     */
  }, {
    key: 'resolveAlias',
    value: _asyncToGenerator(function* (aliasOrTarget) {
      var args = ['targets', '--resolve-alias', aliasOrTarget];
      var result = yield this._runBuckCommandFromProjectRoot(args);
      return result.stdout.trim();
    })

    /**
     * Currently, if `aliasOrTarget` contains a flavor, this will fail.
     *
     * @return Promise resolves to absolute path to output file
     */
  }, {
    key: 'outputFileFor',
    value: _asyncToGenerator(function* (aliasOrTarget) {
      var args = ['targets', '--show-output', aliasOrTarget];
      var result = yield this._runBuckCommandFromProjectRoot(args);
      var stdout = result.stdout.trim();
      if (stdout.indexOf(' ') !== -1) {
        var relativePath = stdout.split(' ')[1];
        return path.resolve(this._rootPath, relativePath);
      } else {
        return null;
      }
    })

    /**
     * Currently, if `aliasOrTarget` contains a flavor, this will fail.
     */
  }, {
    key: 'buildRuleTypeFor',
    value: _asyncToGenerator(function* (aliasOrTarget) {
      var args = ['targets', '--json', aliasOrTarget];
      var result = yield this._runBuckCommandFromProjectRoot(args);
      var json = JSON.parse(result.stdout.trim());
      return json[0]['buck.type'];
    })
  }, {
    key: 'getServerPort',
    value: _asyncToGenerator(function* () {
      var args = ['server', 'status', '--json', '--http-port'];
      var result = yield this._runBuckCommandFromProjectRoot(args);
      var json = JSON.parse(result.stdout);
      return json['http.port'];
    })
  }]);

  return LocalBuckProject;
})(BuckProject);

module.exports = LocalBuckProject;
