// nuclide-clang/lib/LocalClangService.js
var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; desc = parent = getter = undefined; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

/*
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the LICENSE file in
 * the root directory of this source tree.
 */

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { var callNext = step.bind(null, 'next'); var callThrow = step.bind(null, 'throw'); function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(callNext, callThrow); } } callNext(); }); }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

'use babel';

var logger = require('nuclide-logging').getLogger();
var ClangService = require('./ClangService');

/**
 * Mock class for LocalClangService implementation.
 * Since nuclide-clang is not ready to be open sourced, we put the implementation into `fb`
 * subfolder as it won't be published, and the open sourced version will load this
 * mock class as ClangService's local implementation to make sure nothing breaks.
 */

var LocalClangServiceMock = (function (_ClangService) {
  _inherits(LocalClangServiceMock, _ClangService);

  function LocalClangServiceMock() {
    _classCallCheck(this, LocalClangServiceMock);

    _get(Object.getPrototypeOf(LocalClangServiceMock.prototype), 'constructor', this).apply(this, arguments);
  }

  _createClass(LocalClangServiceMock, [{
    key: 'compile',
    value: function compile(src, contents) {
      return Promise.resolve({
        diagnostics: []
      });
    }
  }, {
    key: 'getCompletions',
    value: function getCompletions(src, contents, line, column, tokenStartColumn, prefix) {
      return {
        file: src,
        completions: [],
        line: line,
        column: column,
        prefix: prefix
      };
    }
  }, {
    key: 'getDeclaration',
    value: _asyncToGenerator(function* (src, contents, line, column) {
      return null;
    })
  }, {
    key: 'getDeclarationInfo',
    value: function getDeclarationInfo(src, contents, line, column) {
      return { src: src, line: line, column: column, info: [] };
    }
  }]);

  return LocalClangServiceMock;
})(ClangService);

try {
  var LocalClangService = require('./fb/LocalClangService');
} catch (e) {
  logger.info('Use LocalClangServiceMock.');
  var LocalClangService = LocalClangServiceMock;
}

module.exports = LocalClangService;
