// nuclide-logging/lib/config.js

/*
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the LICENSE file in
 * the root directory of this source tree.
 */

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { var callNext = step.bind(null, 'next'); var callThrow = step.bind(null, 'throw'); function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(callNext, callThrow); } } callNext(); }); }; }

var _require = require('nuclide-commons');

var fsPromise = _require.fsPromise;

var path = require('path');

var USER = require('nuclide-commons').env.USER;

var LOG_FILE_PATH = '/tmp/nuclide-' + USER + '-logs/nuclide.log';

var logDirectory = path.dirname(LOG_FILE_PATH);
var logDirectoryInitialized = false;

module.exports = {
  getDefaultConfig: _asyncToGenerator(function* () {

    if (!logDirectoryInitialized) {
      yield fsPromise.mkdirp(logDirectory);
      logDirectoryInitialized = true;
    }

    return {
      appenders: [{
        'type': 'logLevelFilter',
        'level': 'INFO',
        'appender': {
          type: path.join(__dirname, './consoleAppender')
        }
      }, {
        type: 'dateFile',
        alwaysIncludePattern: true,
        absolute: true,
        filename: LOG_FILE_PATH,
        pattern: '-yyyy-MM-dd'
      }]
    };
  }),

  LOG_FILE_PATH: LOG_FILE_PATH
};
